streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultStateConstrainedBalancedAssignor.java

```diff
@@ -0,0 +1,252 @@
+     * @param tasksToCaughtUpClients map from task IDs to lists of caught-up clients
+     */
+    private void balance(final Map<ID, List<TaskId>> assignment,
+                         final int balanceFactor,
+                         final Map<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients,
+                         final Map<TaskId, List<ID>> tasksToCaughtUpClients) {
+        final List<ID> clients = new ArrayList<>(assignment.keySet());
+        Collections.sort(clients);
+        for (final ID client : clients) {
+            final List<TaskId> source = assignment.get(client);
+            for (final TaskId task : assignedTasksWithoutCaughtUpClients(source, tasksToCaughtUpClients)) {
+                for (final ClientIdAndLag<ID> clientAndLag : statefulTasksToRankedClients.get(task)) {
+                    final List<TaskId> destination = assignment.get(clientAndLag.clientId());
+                    if (source.size() - destination.size() > balanceFactor) {
+                        source.remove(task);
+                        destination.add(task);
+                    }
+                }
+            }
+            for (final TaskId task : assignedTasksWithCaughtUpClients(source, tasksToCaughtUpClients)) {
+                for (final ID caughtUpClient : tasksToCaughtUpClients.get(task)) {
+                    final List<TaskId> destination = assignment.get(caughtUpClient);
+                    if (source.size() - destination.size() > balanceFactor) {
+                        source.remove(task);
+                        destination.add(task);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns a sublist of tasks in the given list that does not have a caught-up client.
+     *
+     * @param tasks list of task IDs
+     * @param tasksToCaughtUpClients map from task IDs to lists of caught-up clients
+     * @return a list of task IDs that does not have a caught-up client
+     */
+    private List<TaskId> assignedTasksWithoutCaughtUpClients(final List<TaskId> tasks,
+                                                             final Map<TaskId, List<ID>> tasksToCaughtUpClients) {
+        final List<TaskId> tasksWithoutCaughtUpClients = new ArrayList<>();
+        for (int i = tasks.size() - 1; i >= 0; --i) {
+            final TaskId task = tasks.get(i);
+            if (!tasksToCaughtUpClients.containsKey(task)) {
+                tasksWithoutCaughtUpClients.add(task);
+            } else {
+                break;
+            }
+        }
+        return Collections.unmodifiableList(tasksWithoutCaughtUpClients);
+    }
+    private List<TaskId> assignedTasksWithCaughtUpClients(final List<TaskId> tasks,
+                                                          final Map<TaskId, List<ID>> tasksToCaughtUpClients) {
+        final List<TaskId> tasksWithCaughtUpClients = new ArrayList<>();
+        for (int i = tasks.size() - 1; i >= 0; --i) {
```

Comment:
I think we need to iterate forwards in this case?

Reply:
Sounds good. The assumption is broken due to swapping tasks around during `balance` without preserving order, is that it?

Reply:
Exactly!